feat: swapper rates without wallet

gomesalexandre opened the pull request:
## Description

This PR allows to get rates without a wallet, under a flag.

Note two ramifications with this solution:

1. Not all Arbitrum Bridge routes are happy without an address, meaning depending on the route, you may not be able to get a rate without a wallet. 
2. As part of the inherent arhitecture chosen here, connecting a wallet will refetch quotes vs. keeping the selected *rate*

```
- [x] plumbing
- [x] types plumbing
- [x] all swappers support priceOrQuote (or similar) logic
  - [x] Arb Bridge
  - [x] ZRX 
  - [x] Li.Fi
  - [x] THOR
  - [x] CoW
- [x] improve UX
  - [x] no invalid balance errors if !hasWallet
  - [x] display full list of assets when no wallet connected, not current subset only
  - [x] figure out swapper loss of input state when connecting a wallet
  - [x] triple check schizz referentially invalidates on wallet connect and fire new quotes instantly
```

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

- closes https://github.com/shapeshift/web/issues/7940
- ~~contributes to~~ ~~closes~~ actually only contributes to https://github.com/shapeshift/web/issues/7941

## Risk

> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

High. This is swapper, and failure to get this right means we could execute a rate instead of a quote, with funds ending in the wrong wallet.

## Testing

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>


### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

- eng. should test with the flag on (flag on and off is irrelevant from a logic perspective, the only thing the flag does is making the trade route public). Product to only regression the regular path with wallet connected i.e flag off.

- With the flag on, ensure that you can now get rates without a wallet connected, for all swappers. Network fees may not be available for all swappers/pairs (e.g can't get them for Portals, nor with THOR UTXOs), which is expected. 
- With the flag on, ensure that going from wallet disconnected and connecting a wallet refetches *quotes* (note, this is a temporary flow for the initial implementation, draft flow in-progress with eng. and product)

- With both the flag on (eng.) and off, ensure that you can still get quotes with a wallet connected
- With both the flag on (eng.) and off, ensure that funds land in the correct account, either manual or custom receive addy

- Ensure that implementation looks sane and runtime guard properly prevent rates from getting executed

### Operations

- [x] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

- Ensure that you can still get quotes for all swappers with a wallet connected
- Ensure that funds land in the correct account, either manual or custom receive addy

## Screenshots (if applicable)

<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->

### Wallet connected 

- Li.Fi

https://jam.dev/c/90e98aaf-2dd9-4715-a0bd-a68a00da51db

- CoW 

https://jam.dev/c/3aeb28fe-34c9-4199-98b0-bb0fdcd00ef8

- ZRX

https://jam.dev/c/11ef2dc7-0321-4dcb-aff6-f479561a1483

- Portals

https://jam.dev/c/690af128-1445-4eae-9c89-511ef630c6cf

- THOR

https://jam.dev/c/0fe2b16a-a87b-452c-af48-c3d601a25d86

### No Wallet

- Li.Fi

<img width="1408" alt="image" src="https://github.com/user-attachments/assets/07e487bc-44da-438e-b738-c13ea83aaa55"

- CoW

<img width="1098" alt="image" src="https://github.com/user-attachments/assets/1a4d9c55-4407-4190-b173-ff5f7153d9ce">

- ZRX

<img width="1073" alt="image" src="https://github.com/user-attachments/assets/b4129e28-ed15-4c21-b5bd-439f9502568b">

- Portals

<img width="1088" alt="image" src="https://github.com/user-attachments/assets/409009fc-7ba0-4d99-9ed9-9c8be460c87e">

- THOR

<img width="1123" alt="image" src="https://github.com/user-attachments/assets/39b0f6c6-c980-4699-b45f-49a4179832bc">

- THOR - long-tail

<img width="1089" alt="image" src="https://github.com/user-attachments/assets/e5ca53b5-884c-4338-bb40-8fc5e2691c5a">





<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->

gomesalexandre commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/getTradeQuote/getTradeQuote.ts:
```diff
@@ -41,9 +45,6 @@ export const getTradeQuoteWithWallet = async (
   return getTradeQuote(
     {
       ...input,
-      sellAsset: input.sellAsset,
-      buyAsset: input.buyAsset,
-      accountNumber: input.accountNumber,
```
Useless redeclaration with same name

gomesalexandre commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/getTradeQuote/getTradeQuote.ts:
```diff
@@ -17,7 +18,10 @@ import { makeSwapErrorRight } from '../../../utils'
 
-export type GetEvmTradeQuoteInputWithWallet = Omit<GetEvmTradeQuoteInput, 'supportsEIP1559'> & {
+export type GetEvmTradeQuoteInputWithWallet = Omit<
+  GetEvmTradeQuoteInputWithWalletInfo,
```
Uses input type with wallet info to get a quote with wallet info - we now split input types between `WithWalletInfo` and `EmptyWalletInfo`

gomesalexandre commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/endpoints.ts:
```diff
@@ -125,6 +125,7 @@ export const arbitrumBridgeApi: SwapperApi = {
       sellAmountIncludingProtocolFeesCryptoBaseUnit,
       sellAsset,
       sendAddress: from,
+      priceOrQuote: 'quote',
```
Added for Arb bridge since super straightforward

gomesalexandre commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/endpoints.ts:
```diff
@@ -114,6 +114,10 @@ export const arbitrumBridgeApi: SwapperApi = {
     const { buyAsset, sellAsset, sellAmountIncludingProtocolFeesCryptoBaseUnit } = step
     const { receiveAddress } = tradeQuote
 
+    // TODO(gomes): when we actually split between TradeQuote and TradeRate in https://github.com/shapeshift/web/issues/7941,
+    // this won't be an issue anymore
```
i.e not for this PR - this is intended to be a minimal rate *exposing* quotes without BIP44 assuming flag on, so runtime guards are added for paranoia

gomesalexandre commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/getTradeQuote/getTradeQuote.ts:
```diff
@@ -78,14 +79,15 @@ export async function getTradeQuote(
   try {
     const swap = await fetchArbitrumBridgeSwap({
-      supportsEIP1559,
+      supportsEIP1559: Boolean(supportsEIP1559),
```
Since it may be undefined in case of no wallet

gomesalexandre commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/utils/fetchArbitrumBridgeSwap.ts:
```diff
@@ -40,13 +41,19 @@ export const fetchArbitrumBridgeSwap = async ({
   allowanceContract: string
   networkFeeCryptoBaseUnit: string
 }> => {
+  if (priceOrQuote === 'quote' && !receiveAddress)
+    throw new Error('receiveAddress is required for Arbitrum Bridge quotes')
+  if (priceOrQuote === 'quote' && !sendAddress)
+    throw new Error('sendAddress is required for Arbitrum Bridge quotes')
```
This shouldn't be an issue at types-level given the generic `priceOrQuote` T added, but safety never hurts

gomesalexandre commented on packages/swapper/src/swappers/ThorchainSwapper/utils/getL1quote.ts:
```diff
@@ -243,12 +254,16 @@ export const getL1quote = async (
-                })
+            const memo = (() => {
+              if (!isConnected) return ''
+              if (!quote.memo) throw new Error('no memo provided')
```
i.e THOR won't return a memo when we don't pass `destinationAddress` (which is valid, and is what allows us to get quotes without a wallet). But if we're connected and we get a quote without memo, then problems.

gomesalexandre commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -280,6 +280,7 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet>
     const { baseFee, fast, average, slow } = await this.providers.http.getPriorityFees()
 
     const computeUnits = await this.providers.http.estimateFees({
+      // @ts-ignore wtf? revert me when fixed
```
^

gomesalexandre commented on .env.base:
```diff
@@ -46,7 +46,8 @@ REACT_APP_FEATURE_FOX_PAGE=false
 REACT_APP_FEATURE_FOX_PAGE_FOX_SECTION=true
 REACT_APP_FEATURE_FOX_PAGE_FOX_FARMING_SECTION=false
-REACT_APP_FEATURE_PHANTOM_WALLET=false
+REACT_APP_FEATURE_PHANTOM_WALLET=true
```
Nah I just cherry-picked https://github.com/shapeshift/web/pull/7964
 in here before it went in, since I noticed this flag went away in a wrong mergefix in one of the FOX Page PRs - should look good after a develop merge in a sec

gomesalexandre replied:
7e2a8b56341d91b37c3148e53f9bfdfe5775bec9

gomesalexandre commented on packages/swapper/src/types.ts:
```diff
@@ -123,41 +123,80 @@ export type BuyAssetBySellIdInput = {
+  hasWallet: false
+}
+
+type CommonTradeInput = CommonTradeQuoteInput | CommonTradeRateInput
```
The intent is making it obvious at types-level those are two different things. This is an input, not a trade quote, so it forces us to pass a very strongly typed rate input or a quote input, not a Frankesteiny mix of both.

I originally went with the optional before going with the type renaming, but that seemed off to me, i.e we could end up with bad states where one would (wrongly) be undefined, while the others were properly defined.

With this types plumbing, if we're dealing with a rate, all the BIP44 bits should *explicitly* be undefined, which allows us to discriminate and will throw types error if we don't. And throwing it did, effectivey had tsc screaming at me and forcing me to have two different paths when dealing with a maybe rate input, which means safety 🙏🏽

gomesalexandre commented on packages/swapper/src/types.ts:
```diff
@@ -229,12 +272,25 @@ export type MultiHopTradeQuote = TradeQuoteBase & {
+  steps: SingleHopTradeQuoteSteps | MultiHopTradeQuoteSteps
+}
+
+export type TradeQuoteOrRate = TradeQuote | TradeRate
```
Not sure there would be much surface area change since in effect, this is really a type rename i.e lsp renamed `TradeQuote` to `TradeQuoteOrRate`.
The main benefit of this *currently* is having a clear-defined type which we can use a `isTradeRate`  type guard for, see an example of consumption here: https://github.com/shapeshift/web/blob/7e2a8b56341d91b37c3148e53f9bfdfe5775bec9/src/lib/tradeExecution.ts#L73-L75

Though keep in mind this PR was originally written with the intent of having a second PR where we do clear rates and quotes firing, i.e rates at input time, then when at confirm time with a wallet, a second quote which is a trade quote, see this copypasta from engineering DM below.

As mentioned in that message, there are many implications to that and given this already does the do functionally vs. keeping rates in memory when connecting a wallet, I'm not sure if that's realistic for us to go with this functional split in the end. Although I believe there's nothing wrong keeping this, as this improves naming, does a clear types distinction, and makes us safer? 

```
On the finish line for https://github.com/shapeshift/web/pull/7939, currently regression testing it before opening and would like more brains on this for Monday producty things/Wednesday gangles.

tl;dr:

We don't have to proper split between quotes and rates, a quote is just an augmented rate with receiveAddress (types, including input types have been revamped toreflect that)
Most of the magic happens with a combination of types and runtime safety, as well as input referential invalidation:
Types and runtime safety ensure that we don't execute a trade rate, but can get one without addresses (from/to), with checks at getUnsignedTx time to ensure we don't build a payload without addies/bip44 stuff
Referential invalidation ensures that as soon as you connect a wallet, quotes refetch

In the end, this is much more straightforward as opposed to properly splitting between quotes (pre-execution time) and rates (input time), as all swappers have many implications re: keeping quotes gotten at input time at the moment. If we did this, we would also have to ensure that we don't end up with too big of rates variations, and reflect said variations to users if any.

Works pretty nicely as-is, and input is persisted, however, as per design, quotes refetch on wallet connection, which is actually part of the safety model here. This means that quotes aren't, and as a result, nor is the selected quote

https://jam.dev/c/b7a20304-e871-4ea8-a772-3b30ba48d756 
https://jam.dev/c/e7c07d97-ce8f-4c34-8578-f86296fe2dc8

The q. here is whether we really want to go with another swapper architectural on top of this PR, to shave a few seconds getting quotes again on wallet connect which is going to be a ~3 of sorts and a high-risk once again, or if we're happy with this flow. 
```

gomesalexandre replied:
After discussing this one out in DM, removed the split for `TradeQuote` (`TradeQuoteOrRate`), though kept for `TradeQuoteInput` for safety: https://github.com/shapeshift/web/pull/7939/commits/299c8a403c81dba47853c027eb7b1b63a891754c

gomesalexandre commented on src/Routes/Routes.tsx:
```diff
@@ -98,7 +98,7 @@ export const Routes = memo(() => {
   // However, some routes are unstable and should actually re-render when location.pathname reference changes
   // Which happens both on actual route change, or when the same route gets pushed
-  const unstableRoutes = useMemo(() => ['/trade'], [])
+  const unstableRoutes = useMemo(() => [], [])
```
Ab-solutely, we can't have that one anymore, else this will produce bugs when connecting a wallet (i.e this will re-render the trade route,  and given we use some random key to make the trade route unstable, this will reset the input).

For historical context, the goal was to *actually* reset the swapper when clicking "Trade" in the side nav, for a better UX, but we can't do that anymore.

gomesalexandre replied:
de59ffb1ee3c8562403e2f32e899f311e5dfb1e1

gomesalexandre commented on src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useTradeExecution.tsx:
```diff
@@ -266,6 +274,9 @@ export const useTradeExecution = (
               } as evm.BuildCustomTxInput)
 
+              if (!receiverAddress) throw Error('Missing receiver address')
+              if (!from) throw Error('Missing sender address')
```
`from` wouldn't be caught by runtime checks, since it's not part of `TradeQuote` (now `TradeQuoteOrRate`) but gotten with `adapter.getAddress()` in place, it's more of a paranoia than anything else, we should really not end up here:
https://github.com/shapeshift/web/blob/ddb90afd0214a90361ee646964575ea9dd101f3f/src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useTradeExecution.tsx#L260

However, `receiverAddress` *is* indeed a runtime check that was made possible thanks to types rehaul:

<img width="710" alt="image" src="https://github.com/user-attachments/assets/8523b02e-d5fa-4e99-9f82-df1cb428e2a3">

@woodenfurniture not sure why we would remove the types change, as we wouldn't have gotten tsc screaming at us and forcing us to narrow this in the first place if we did so? Sure, it's *now* guarded at runtime, but only for now in this diff, if we were to remove the types change, we wouldn't have the type safety anymore and would have no safety under our belt in case we touch e.g this file?

gomesalexandre commented on src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useTradeExecution.tsx:
```diff
@@ -318,6 +334,8 @@ export const useTradeExecution = (
           return
         }
         case CHAIN_NAMESPACE.CosmosSdk: {
+          if (isTradeRate(tradeQuote)) throw new Error('Cannot execute a trade rate')
```
See https://github.com/shapeshift/web/pull/7939#discussion_r1808384086 on why we probably don't want to remove type safety 🙏🏽

gomesalexandre replied:
https://github.com/shapeshift/web/pull/7939/commits/299c8a403c81dba47853c027eb7b1b63a891754c

kaladinlight commented on packages/swapper/src/swappers/ZrxSwapper/endpoints.ts:
```diff
@@ -42,6 +42,8 @@ export const zrxApi: SwapperApi = {
     const { affiliateBps, receiveAddress, slippageTolerancePercentageDecimal, steps } = tradeQuote
     const { buyAsset, sellAsset, sellAmountIncludingProtocolFeesCryptoBaseUnit } = steps[0]
 
+    if (!receiveAddress) throw new Error('receiveAddress is required for ZrxSwapper quotes')
```
just noticed, we are kind of missusing receive address for zrx. cross account is not supported and really should be `sendAddress` for all intents and purposes as used for `takerAddress` in the price/quote request. also just generally speaking it feels like it would make more sense to use `hasWallet` as the primary discriminator to make it more clear the intention imo.

kaladinlight replied:
just noticed this is from `TradeQuote` vs `GetTradeQuoteInput` which does not include `hasWallet`. in this case I suppose it makes sense enough

kaladinlight replied:
can use `isExecutableTradeQuote` though.

gomesalexandre replied:
> just noticed this is from `TradeQuote` vs `GetTradeQuoteInput` which does not include `hasWallet`. in this case I suppose it makes sense enough

Precisely, while this looks like a big diff at first glance, the intent is for a minimal changeset here and not try and change the underlying implementations of each swapper unless required to make it work without a wallet.

Agreed ZRX (and well, not only, CoW would be too) is weird terminology-wise, which is something I've noticed while working in this domain, feels weird that `receiveAddress` is the discriminator... but it's really what we use both for send side (if cross-account not supported, else receive address must be different) *and* receive side across the board, which is probably something we want to tackle in a follow-up 🙏🏽 

And nice catch on `isExecutableTradeQuote`! 486607c07f

kaladinlight commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/endpoints.ts:
```diff
@@ -112,6 +112,9 @@ export const arbitrumBridgeApi: SwapperApi = {
 
     const { buyAsset, sellAsset, sellAmountIncludingProtocolFeesCryptoBaseUnit } = step
+
+    if (!isExecutableTradeQuote(tradeQuote)) throw new Error('Cannot execute a trade rate')
```
Maybe something more generic, could even mention no receive address or no wallet verbiage for more context. Reason I comment is no being able to execute a trade rate wouldn't really be accurate once rate and quote are split. We will always be sending a trade rate "quote" at that point and it will be able to be executed given we have a receive address or more accurately a wallet is connected.
```suggestion
    if (!isExecutableTradeQuote(tradeQuote)) throw new Error('Unable to execute trade')
```

gomesalexandre replied:
As discussed, not 100% set on the split yet, but regardless, a more generic error does make sense! 5cb17e941a

kaladinlight commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/getTradeQuote/getTradeQuote.ts:
```diff
@@ -34,20 +35,19 @@ export const isArbitrumBridgeTradeQuote = (
     deps,
   )
+
+  return quote as Result<ArbitrumBridgeTradeQuote, SwapErrorRight>
```
nitpick - no need for the casted return that I'm seeing

gomesalexandre replied:
Nice catch, leftover from initial types plumbing (when quotes and rates where split at type-level): f79211f001

kaladinlight commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/getTradeQuote/getTradeQuote.ts:
```diff
@@ -78,14 +78,15 @@ export async function getTradeQuote(
       buyAsset,
       sellAmountIncludingProtocolFeesCryptoBaseUnit,
       sellAsset,
       sendAddress: sendAddress ?? '',
```
```suggestion
      sendAddress,
```

gomesalexandre replied:
8d0ffb1c60

kaladinlight commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/getTradeQuote/getTradeQuote.ts:
```diff
@@ -78,14 +78,15 @@ export async function getTradeQuote(
       sendAddress: sendAddress ?? '',
       receiveAddress,
       assertGetEvmChainAdapter,
+      priceOrQuote: 'price',
```
just making a note that if we have a wallet (ie. send/receive address), this type would actually reflect `quote` based on the type definition. also price is broken for arbitrum bridge as we cannot estimate gas without an address. suggestion to either omit gas estimation to show the 1:1 rate, or leverage the placeholder estimation address strategy we took at one point.

gomesalexandre replied:
Added more safety as well as price vs. quote when a wallet is connected in f8ef5825bd

gomesalexandre replied:
https://github.com/shapeshift/web/issues/7976 for follow-up

kaladinlight commented:
I am unable to swap assets without a wallet connected and saw this state with the no wallet address on no wallet connect but haven't been able to recreate yet (might be just a dev env issue)
![image](https://github.com/user-attachments/assets/7bf279cb-411d-4556-a3b0-82264c344cc2)

kaladinlight commented on packages/swapper/src/swappers/CowSwapper/endpoints.ts:
```diff
@@ -78,6 +79,8 @@ export const cowApi: SwapperApi = {
       ),
     } = tradeQuote
 
+    if (receiveAddress === zeroAddress) throw new Error('No receive address found in quote')
```
we don't return `zeroAddress` on the cowswap quote. can use `isExecutableTrade` for safety.

gomesalexandre replied:
Very nice catch! a7007bab60

kaladinlight commented on packages/swapper/src/swappers/LifiSwapper/endpoints.ts:
```diff
@@ -98,7 +98,13 @@ export const lifiApi: SwapperApi = {
-    if (to === undefined || value === undefined || data === undefined || gasLimit === undefined) {
+    if (
+      to === undefined ||
+      from === undefined ||
```
from is defined as `string` in the args. though feels like we can add the `isExecutableTradeQuote` here as well

gomesalexandre replied:
🧠  4c316cb35f

kaladinlight commented on packages/swapper/src/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts:
```diff
@@ -38,8 +39,18 @@ export async function getCowSwapTradeQuote(
   } = input
 
+  // This should never happen, and we have additional checks at execution time re: receiveAddress === zeroAddress but...
+  if (hasWallet && !(receiveAddress && accountNumber !== undefined))
```
do we want to add this safety check to arbitrum and cow as well?

gomesalexandre replied:
- Added to Arbitrum in f8ef5825bd0 
- Additional safety for Li.Fi in cc75a5d2bd
- Additional safety for Arb in 7a0e3f7ab9

kaladinlight commented on packages/swapper/src/swappers/OneInchSwapper/endpoints.ts:
```diff
@@ -53,6 +53,10 @@ export const oneInchApi: SwapperApi = {
 
+    // TODO(gomes): when we actually split between TradeQuote and TradeRate in https://github.com/shapeshift/web/issues/7941,
+    // this won't be an issue anymore
+    if (!receiveAddress) throw new Error('receiveAddress is required for OneInchSwapper quotes')
```
should we just delete this swapper since we aren't using anymore? also `isExecutableTradeQuote`

gomesalexandre replied:
Always in for some bloodbath. 8890a9f4ff4e7852f6e8703ecb1fed681ed691a2

kaladinlight commented on packages/swapper/src/swappers/PortalsSwapper/endpoints.ts:
```diff
@@ -49,6 +49,8 @@ export const portalsApi: SwapperApi = {
     const { affiliateBps, slippageTolerancePercentageDecimal, steps } = tradeQuote
     const { buyAsset, sellAsset, sellAmountIncludingProtocolFeesCryptoBaseUnit } = steps[0]
 
+    if (!from) throw new Error('Cannot execute a Portals trade without a from address')
```
`isExecutableTradeQuote`?

kaladinlight commented on packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts:
```diff
@@ -109,8 +112,9 @@ export async function getPortalsTradeQuote(
-      sender: sendAddress,
+    const portalsTradeQuoteResponse = await fetchPortalsTradeOrder({
+      hasWallet,
+      sender: hasWallet ? sendAddress! : undefined,
```
all of these don't need the ternary which is just setting it to basically what they would be defined as anyways.
```suggestion
      sender: sendAddress,
```

gomesalexandre replied:
Another one of trade vs. quote split leftovers! 269a24149f

kaladinlight commented on packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts:
```diff
@@ -128,7 +132,8 @@ export async function getPortalsTradeQuote(
 
```
comment for the above line. lots of spam

kaladinlight commented on packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts:
```diff
@@ -181,24 +189,55 @@ export async function getPortalsTradeQuote(
         validate: false,
         swapperConfig,
+      }).catch(e => {
+        console.info('failed to get Portals quote with validation disabled', e)
```
lots of spam

gomesalexandre replied:
43fe29b9a8

kaladinlight commented on packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts:
```diff
@@ -109,8 +112,9 @@ export async function getPortalsTradeQuote(
     // Attempt fetching a quote with validation enabled to leverage upstream gasLimit estimate
-    const portalsTradeOrderResponse = await fetchPortalsTradeOrder({
-      sender: sendAddress,
+    const portalsTradeQuoteResponse = await fetchPortalsTradeOrder({
```
seems like we could be smarter with these quotes if we don't have a wallet to save two guranteed failed requests. could probably wait until part 2 follow up though.

gomesalexandre replied:
43fe29b9a8

kaladinlight commented on packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts:
```diff
@@ -181,24 +189,55 @@ export async function getPortalsTradeQuote(
       })
 
-      if (dummyOrderResponse?.maybeGasLimit)
+      if (quoteEstimateResponse && !order) {
```
I am never seeing this condition hit which would negate the need for all of the response payload checks below. Curious what flow ended up here. I am getting an order both with and without a wallet.

gomesalexandre replied:
Cleaned up logic in https://github.com/shapeshift/web/commit/43fe29b9a829a28688c10c66e1cd68c5cdabb6e9 with a clear branch in case of `!hasWallet`, should be much cleaner now!

kaladinlight commented on packages/swapper/src/swappers/ThorchainSwapper/endpoints.ts:
```diff
@@ -47,7 +48,7 @@ const deductOutboundRuneFee = (fee: string): string => {
 export const thorchainApi: SwapperApi = {
   getTradeQuote: async (
-    input: GetTradeQuoteInput,
+    input: GetTradeQuoteInput | GetTradeRateInput,
```
As it stands, `GetTradeQuoteInput` also contains all of the trade rate types.

gomesalexandre replied:
Not exactly, one has specific fields explicitly set to undefined which forces us to be explicit in checks vs. adding `| undefined` in base type

kaladinlight replied:
Correct, I see the distinction between the types, but `GetTradeQuoteInput` is comprised of a union of rate and quote for each type (utxo, evm, cosmossdk) already.

example:
```
export type GetUtxoTradeQuoteInput = GetUtxoTradeQuoteWithWallet | GetUtxoTradeRateInput

export type GetTradeQuoteInput =
  | GetUtxoTradeQuoteInput
  | GetEvmTradeQuoteInput
  | GetCosmosSdkTradeQuoteInput
```

problem is we are now double levaraging functions so naming has become a bit weird. ideally it would be something like `getTradeRateOrQuote(input: GetTradeQuoteInput | GetTradeRateInput)` maybe depending on how we finish building this out. And then `GetTradeQuoteInput` would only contain quote args and `GetTradeRateInput` would only contain rate args.

gomesalexandre replied:
Oh indeed, that union was a smell! https://github.com/shapeshift/web/pull/7939/commits/ee753576f085713474088c70b057ff7c67f10352

kaladinlight commented on packages/swapper/src/swappers/ThorchainSwapper/endpoints.ts:
```diff
@@ -84,6 +85,9 @@ export const thorchainApi: SwapperApi = {
     } = tradeQuote as ThorEvmTradeQuote
     const { sellAmountIncludingProtocolFeesCryptoBaseUnit, sellAsset } = steps[0]
 
+    if (!from) throw new Error('Cannot execute Tx without a from address')
```
`isExecutableTradeQuote`?

gomesalexandre replied:
2e58d5b8b8

kaladinlight commented on packages/swapper/src/swappers/ThorchainSwapper/endpoints.ts:
```diff
@@ -212,6 +216,9 @@ export const thorchainApi: SwapperApi = {
     const { accountNumber, sellAmountIncludingProtocolFeesCryptoBaseUnit, sellAsset, feeData } =
       steps[0]
 
+    if (accountNumber === undefined) throw new Error('Cannot execute Tx without an accountNumber')
```
`isExecutableTradeQuote`?

gomesalexandre replied:
2e58d5b8b8

kaladinlight commented on packages/swapper/src/swappers/ThorchainSwapper/endpoints.ts:
```diff
@@ -243,7 +250,11 @@ export const thorchainApi: SwapperApi = {
+    const { accountNumber, sellAmountIncludingProtocolFeesCryptoBaseUnit, sellAsset, feeData } =
+      steps[0]
+
+    if (accountNumber === undefined) throw new Error('Cannot execute Tx without an accountNumber')
```
`isExecutableTradeQuote`?

gomesalexandre replied:
2e58d5b8b8

gomesalexandre commented on packages/swapper/src/swappers/ZrxSwapper/utils/fetchFromZrx.ts:
```diff
@@ -17,7 +17,7 @@ export type FetchFromZrxArgs<T extends 'price' | 'quote'> = {
   sellAsset: Asset
   sellAmountIncludingProtocolFeesCryptoBaseUnit: string
-  receiveAddress: string
+  receiveAddress: T extends 'price' ? string | undefined : string
```
Prefer not to tackle in this PR as unrelated, but added to TODOs 🙏🏽

kaladinlight replied:
yep, makes sense, thanks!

kaladinlight commented on src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useTradeExecution.tsx:
```diff
@@ -208,6 +213,7 @@ export const useTradeExecution = (
       const stepBuyAssetAssetId = hop.buyAsset.assetId
 
       if (swapperName === SwapperName.CowSwap) {
+        if (!isExecutableTradeQuote(tradeQuote)) throw new Error('Unable to execute trade')
```
can just do this single check at the top and then we are guaranteed to have the receive address for execute and don't need the extra runtime checks in these if conditions

gomesalexandre replied:
Added in 32e54ca544 - note not at the very very top of the function so we ensure all event handlers are still present

kaladinlight commented on src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useTradeExecution.tsx:
```diff
@@ -248,6 +254,8 @@ export const useTradeExecution = (
 
       switch (stepSellAssetChainNamespace) {
         case CHAIN_NAMESPACE.Evm: {
+          if (isExecutableTradeQuote(tradeQuote)) throw new Error('Unable to execute trade')
```
also this and the cosmos sdk condition are inverted and wrong

gomesalexandre replied:
Consolidated in 32e54ca544 - note not at the very very top of the function so we ensure all event handlers are still present

kaladinlight commented on src/components/MultiHopTrade/hooks/useGetTradeQuotes/getTradeQuoteInput.ts:
```diff
@@ -34,14 +25,30 @@ export type GetTradeQuoteInputArgs = {
   isSnapInstalled?: boolean
   pubKey?: string | undefined
-}
+} & (
+  | {
+      receiveAccountNumber?: number
+      receiveAddress: string
+      sellAccountNumber: number
+      wallet: HDWallet
+      hasWallet: true
+    }
+  | {
+      receiveAccountNumber?: number
+      receiveAddress: string | undefined
+      sellAccountNumber: number | undefined
+      wallet: HDWallet | undefined
+      hasWallet: false
+    }
+)
```
this union doesn't seem important at this level imo. we are just passing through the args either defined or not and the swappers handle the discrimination accordingly.

gomesalexandre replied:
@kaladinlight don't we feel like it's better to have two clearly separate types, regardless of the discrimination being handled at runtime or not?

gomesalexandre replied:
Actually reading https://github.com/shapeshift/web/pull/7939#discussion_r1809643839, this does make sense in terms of ergonomics, taking a look now

gomesalexandre replied:
6f13836f3e

kaladinlight commented on src/components/MultiHopTrade/hooks/useGetTradeQuotes/getTradeQuoteInput.ts:
```diff
@@ -70,38 +77,89 @@ export const getTradeQuoteInput = async ({
+      if (wallet && receiveAccountNumber === undefined)
+        throw new Error('missing receiveAccountNumber')
+
+      return hasWallet && receiveAccountNumber !== undefined
+        ? {
+            ...tradeQuoteInputCommonArgs,
+            chainId: sellAsset.chainId as EvmChainId,
+            hasWallet,
+            supportsEIP1559: supportsEIP1559!,
+            receiveAddress,
+            accountNumber: sellAccountNumber,
+            ...(sendAddress ? { sendAddress } : {}),
+            receiveAccountNumber,
+          }
+        : {
+            ...tradeQuoteInputCommonArgs,
+            chainId: sellAsset.chainId as EvmChainId,
+            hasWallet: false,
+            supportsEIP1559: undefined,
+            receiveAddress: undefined,
+            accountNumber: undefined,
+            sendAddress: undefined,
+            receiveAccountNumber: undefined,
+          }
```
would save all of the ternaries too

gomesalexandre replied:
6f13836f3e

kaladinlight commented on src/components/MultiHopTrade/hooks/useGetTradeQuotes/useGetTradeQuotes.tsx:
```diff
@@ -222,24 +231,57 @@ export const useGetTradeQuotes = () => {
-      })
-
-      setTradeQuoteInput(updatedTradeQuoteInput)
+      if (hasWallet) {
```
feels like we don't need this if condition either. just pass the args to the swapper where all validation and discrimination is handled.

kaladinlight commented on src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx:
```diff
@@ -216,6 +217,8 @@ export const TradeInput = ({ isCompact, tradeInputRef, onChangeTab }: TradeInput
       if (!tradeQuoteStep) throw Error('missing tradeQuoteStep')
       if (!activeQuote) throw Error('missing activeQuote')
 
+      if (!isExecutableTradeQuote(activeQuote)) throw new Error('Cannot submit a trade rate')
```
same generic `Unable to execute trade` error?

gomesalexandre replied:
66415441ab

kaladinlight commented on packages/swapper/src/swappers/ZrxSwapper/utils/fetchFromZrx.ts:
```diff
@@ -65,7 +65,8 @@ export const fetchFromZrx = async <T extends 'price' | 'quote'>({
       affiliateAddress: AFFILIATE_ADDRESS, // Used for 0x analytics
-      skipValidation: priceOrQuote === 'price', // don't validate allowances for price queries
+      // Always skip validation, so that we can get a quote even if no wallet is connected
+      skipValidation: true,
```
if we skip this on quote, do we still get back `estimatedGas`? I would think validation is the simulation of the tx that would also provide that, but I could be mistaken. We do use and need `estimatedGas` back from the quote as we were getting failed trades without it.

gomesalexandre replied:
Regular ZRX quotes have been disabled on all envs as of yesterday (permit2 enabled),and we need a bit of different parsing there which has been done in https://github.com/shapeshift/web/pull/7939/commits/db607784b4d037d022afaf39ab71e79fa571c104 but yes, price response for regular ZRX quotes include `estimatedGas` with skipped validation! https://paw.pt/jabNXKeV

<img width="1728" alt="image" src="https://github.com/user-attachments/assets/9f568a4c-2b5c-46c2-a84c-25d0a8a79005">

And with permit2 enabled: https://paw.pt/jabP3abI

<img width="1725" alt="image" src="https://github.com/user-attachments/assets/56675528-bc3a-4b50-8c84-fbc09e7281ab">

kaladinlight replied:
Just making a note that the price `estimatedGas` is most likely a rough estimation versus an actual estimation with `validation`. I remember seeing a diff between the two states on either 0x or Lifi (can't exactly recall which). Just want to validate this change since it is changing behavior from prod and we can still skip validation on price and use it on quote as it was.

gomesalexandre commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/utils/constants.ts:
```diff
@@ -7,3 +8,26 @@ export const ARBITRUM_BRIDGE_SUPPORTED_CHAIN_IDS: SupportedChainIds = {
+// Broad estimate calculated by looking at a couple of different ERC-20 deposits
+// https://github.com/OffchainLabs/arbitrum-token-bridge/blob/d17c88ef3eef3f4ffc61a04d34d50406039f045d/packages/arb-token-bridge-ui/src/util/TokenDepositUtils.ts#L45-L51
+export const fallbackErc20DepositGasLimit = bn(240_000)
+// Broad estimate calculated by looking at a couple of different ERC-20 withdraws
+// https://arbiscan.io/tx/0xf27939d382abcb0cce5c202489db457a6cc0d0dd8062468543400c3bf321148f
+// https://arbiscan.io/tx/0xa4639374806ecc1e9de7beafbab2567c078483b84b708e862bfbd84fbc2fc1da
+// https://arbiscan.io/tx/0xb6c3bce7999b2ae4bbe51a64bc7ab370d21ce9bf6807b805239acbf4c244a6db
+// https://arbiscan.io/tx/0x878aa224a65d831c931192550b3d8fe114fa81660c1af8369c0e2ebea682dd5b
+// https://arbiscan.io/tx/0xf293bd64f9dabddaffc4c8b97f2a602d4e9f77565f5d24018a0f70a95c1ecd38
```
Discussed this one out, those are newly added examples for all new cases (Arb Bridge repo only has that for ERC20 deposits)

kaladinlight commented on src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useTradeExecution.tsx:
```diff
@@ -317,6 +328,9 @@ export const useTradeExecution = (
                 wallet,
               })
 
+              if (!receiverAddress) throw Error('Missing receiver address')
```
nitpick - we are guaranteed the `receiverAddress` now with the top level `isExecutableTrade` and can omit the runtime sender/from address checks as well

gomesalexandre replied:
🧠  
- removed for `receiverAddress` in b1baf3d8e9
- removed `from` check in 5b02fc8756 - this one was useless either way, `adapter.getAddress()` would throw and can't return falsy

kaladinlight commented on src/components/MultiHopTrade/hooks/useGetTradeQuotes/getTradeQuoteInput.ts:
```diff
@@ -70,41 +67,75 @@ export const getTradeQuoteInput = async ({
     case CHAIN_NAMESPACE.Evm: {
-      const supportsEIP1559 = supportsETH(wallet) && (await wallet.ethSupportsEIP1559())
+      const supportsEIP1559 =
+        hasWallet && wallet && supportsETH(wallet) && (await wallet.ethSupportsEIP1559())
```
nitpick
```suggestion
        supportsETH(wallet) && (await wallet.ethSupportsEIP1559())
```

gomesalexandre replied:
0de700b10c - kept in other places since relevant to current types

woodenfurniture commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/utils/fetchArbitrumBridgeSwap.ts:
```diff
@@ -67,44 +75,98 @@ export const fetchArbitrumBridgeSwap = async ({
+              })
+          : undefined
+
+      const networkFeeCryptoBaseUnit = await (async () => {
+        // Fallback fees
+        if (!maybeRequest) {
+          const { average } = await adapter.getGasFeeData()
+          const { gasPrice, maxFeePerGas } = average
+
+          // eip1559 fees
+          if (supportsEIP1559 && maxFeePerGas) {
+            return fallbackEthDepositGasLimit.times(maxFeePerGas).toFixed()
+          }
+
+          // legacy fees
+          return fallbackEthDepositGasLimit.times(gasPrice).toFixed()
+        }
+
+        // Actual fees
+        const feeData = await evm.getFees({
+          adapter,
+          data: maybeRequest.txRequest.data.toString(),
+          to: maybeRequest.txRequest.to,
+          value: maybeRequest.txRequest.value.toString(),
+          from: maybeRequest.txRequest.from,
+          supportsEIP1559,
+        })
+
+        return feeData.networkFeeCryptoBaseUnit
+      })()
```
Nitpick: can we lift this whole IIFE into a util? 3 near-dupes of this would benefit

gomesalexandre replied:
Will do in curry PR 🙏🏽 https://github.com/shapeshift/web/pull/7986

gomesalexandre replied:
gm https://github.com/shapeshift/web/pull/7986/commits/4b1a8141d8b34077f2e9e24828751fccd14f0f75

woodenfurniture commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/utils/fetchArbitrumBridgeSwap.ts:
```diff
@@ -67,44 +75,98 @@ export const fetchArbitrumBridgeSwap = async ({
+              })
+          : undefined
+
+      const networkFeeCryptoBaseUnit = await (async () => {
+        // Fallback fees
+        if (!maybeRequest) {
+          const { average } = await adapter.getGasFeeData()
+          const { gasPrice, maxFeePerGas } = average
+
+          // eip1559 fees
+          if (supportsEIP1559 && maxFeePerGas) {
+            return fallbackEthWithdrawGasLimit.times(maxFeePerGas).toFixed()
+          }
+
+          // legacy fees
+          return fallbackEthWithdrawGasLimit.times(gasPrice).toFixed()
+        }
+
+        // Actual fees
+        const feeData = await evm.getFees({
+          adapter,
+          data: maybeRequest.txRequest.data.toString(),
+          to: maybeRequest.txRequest.to,
+          value: maybeRequest.txRequest.value.toString(),
+          from: maybeRequest.txRequest.from,
+          supportsEIP1559,
+        })
+
+        return feeData.networkFeeCryptoBaseUnit
+      })()
```
Nitpick: here too - can we lift this whole IIFE into a util?

gomesalexandre replied:
Added to https://github.com/shapeshift/web/pull/7986 TODOs

gomesalexandre replied:
gm https://github.com/shapeshift/web/pull/7986/commits/4b1a8141d8b34077f2e9e24828751fccd14f0f75

woodenfurniture commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/utils/fetchArbitrumBridgeSwap.ts:
```diff
@@ -168,23 +233,49 @@ export const fetchArbitrumBridgeSwap = async ({
+              })
+          : undefined
+
+      const networkFeeCryptoBaseUnit = await (async () => {
+        // Fallback fees
+        if (!maybeRequest) {
+          const { average } = await adapter.getGasFeeData()
+          const { gasPrice, maxFeePerGas } = average
+
+          // eip1559 fees
+          if (supportsEIP1559 && maxFeePerGas) {
+            return fallbackErc20WithdrawGasLimit.times(maxFeePerGas).toFixed()
+          }
+
+          // legacy fees
+          return fallbackErc20WithdrawGasLimit.times(gasPrice).toFixed()
+        }
+
+        // Actual fees
+        const feeData = await evm.getFees({
+          adapter,
+          data: maybeRequest.txRequest.data.toString(),
+          to: maybeRequest.txRequest.to,
+          value: maybeRequest.txRequest.value.toString(),
+          from: maybeRequest.txRequest.from,
+          supportsEIP1559,
+        })
+
+        return feeData.networkFeeCryptoBaseUnit
+      })()
```
Nitpick: here too - can we lift this whole IIFE into a util?

gomesalexandre replied:
Added to https://github.com/shapeshift/web/pull/7986 TODOs

gomesalexandre replied:
gm https://github.com/shapeshift/web/pull/7986/commits/4b1a8141d8b34077f2e9e24828751fccd14f0f75

woodenfurniture commented on packages/swapper/src/swappers/PortalsSwapper/utils/fetchPortalsTradeOrder.ts:
```diff
@@ -81,12 +82,15 @@ export const fetchPortalsTradeOrder = async ({
   const params = new URLSearchParams({
     partner,
-    sender,
+    ...(sender ? { sender } : {}),
```
remark: `{ [key]: undefined }` not being valid here is why i hate (type|java)script

gomesalexandre replied:
srsly

woodenfurniture commented on packages/swapper/src/swappers/PortalsSwapper/utils/fetchPortalsTradeOrder.ts:
```diff
@@ -119,11 +123,14 @@ export const fetchPortalsTradeEstimate = async ({
+
   const params = new URLSearchParams({
-    sender,
+    ...(sender ? { sender } : {}),
```
remark: `{ [key]: undefined }` not being valid here is why i hate (type|java)script

woodenfurniture commented on packages/swapper/src/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts:
```diff
@@ -225,7 +226,80 @@ async function _getZrxPermit2TradeQuote(
+  // a *rate* quote, not a *quote* quote on quote quote.
+  if (!hasWallet) {
+    const maybeZrxPriceResponse = await fetchFromZrxPermit2({
+      priceOrQuote: 'price',
```
👀

gomesalexandre replied:
it's happening!

NeOMakinG replied:
quote quote quote

![image](https://github.com/user-attachments/assets/9890a011-036a-4c68-8f5a-4d6cfc4b1e5a)

woodenfurniture commented on packages/swapper/src/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts:
```diff
@@ -225,7 +226,80 @@ async function _getZrxPermit2TradeQuote(
+      return Ok({
+        id: uuid(),
+        receiveAddress,
+        potentialAffiliateBps,
+        affiliateBps,
+        // Slippage protection is only provided for specific pairs.
+        // If slippage protection is not provided, assume a no slippage limit.
+        // If slippage protection is provided, return the limit instead of the estimated slippage.
+        // https://0x.org/docs/0x-swap-api/api-references/get-swap-v1-quote
+        slippageTolerancePercentageDecimal,
+        rate,
+        steps: [
+          {
+            estimatedExecutionTimeMs: undefined,
+            // We don't care about this - this is a rate, and if we really wanted to, we know the permit2 allowance target
+            allowanceContract: undefined,
+            buyAsset,
+            sellAsset,
+            accountNumber,
+            rate,
+            feeData: {
+              protocolFees: {},
+              networkFeeCryptoBaseUnit, // L1 fee added inside of evm.calcNetworkFeeCryptoBaseUnit
+            },
+            buyAmountBeforeFeesCryptoBaseUnit,
+            buyAmountAfterFeesCryptoBaseUnit,
+            sellAmountIncludingProtocolFeesCryptoBaseUnit,
+            source: SwapperName.Zrx,
+          },
+        ] as unknown as SingleHopTradeQuoteSteps,
+      })
+    } catch (err) {
+      return Err(
+        makeSwapErrorRight({
+          message: 'failed to get fee data',
+          cause: err,
+          code: TradeQuoteError.NetworkFeeEstimationFailed,
+        }),
+      )
+    }
```
preferably blocking: nothing to catch 😅

```suggestion
    const networkFeeCryptoBaseUnit = totalNetworkFee
    return Ok({
      id: uuid(),
      receiveAddress,
      potentialAffiliateBps,
      affiliateBps,
      // Slippage protection is only provided for specific pairs.
      // If slippage protection is not provided, assume a no slippage limit.
      // If slippage protection is provided, return the limit instead of the estimated slippage.
      // https://0x.org/docs/0x-swap-api/api-references/get-swap-v1-quote
      slippageTolerancePercentageDecimal,
      rate,
      steps: [
        {
          estimatedExecutionTimeMs: undefined,
          // We don't care about this - this is a rate, and if we really wanted to, we know the permit2 allowance target
          allowanceContract: undefined,
          buyAsset,
          sellAsset,
          accountNumber,
          rate,
          feeData: {
            protocolFees: {},
            networkFeeCryptoBaseUnit, // L1 fee added inside of evm.calcNetworkFeeCryptoBaseUnit
          },
          buyAmountBeforeFeesCryptoBaseUnit,
          buyAmountAfterFeesCryptoBaseUnit,
          sellAmountIncludingProtocolFeesCryptoBaseUnit,
          source: SwapperName.Zrx,
        },
      ] as unknown as SingleHopTradeQuoteSteps,
    })
```

gomesalexandre replied:
Will tackle in https://github.com/shapeshift/web/pull/7986 too which is an effective cleanup PR 🙏🏽

gomesalexandre replied:
gm 12b8327637

woodenfurniture commented on packages/swapper/src/utils.ts:
```diff
@@ -285,3 +286,6 @@ export const getRate = ({
 }
+
+export const isExecutableTradeQuote = (quote: TradeQuote): quote is TradeQuoteWithReceiveAddress =>
+  !!quote.receiveAddress
```
🎉

woodenfurniture commented on src/Routes/Routes.tsx:
```diff
@@ -113,18 +105,14 @@ export const Routes = memo(() => {
-            path={route.path}
-            hasWallet={hasWallet}
-          >
+          <PrivateRoute key={'privateRoute'} path={route.path} hasWallet={hasWallet}>
```
q: does having multiple of the same key cause react warning spew?

gomesalexandre replied:
Surprisingly not in the context of react-router `<Route />`s, no! 
<img width="597" alt="image" src="https://github.com/user-attachments/assets/7bf0554c-fbd2-4054-a3eb-2602106fdb74">

For historical context, back in the early days of the app we had a *lot*, and I mean a lot of bugs and perf issues related to missing keys in `<Route />`s so the same was added for all, iirc we couldn't use a unique one for each since that would cause re-renders. I could absolutely be wrong here but a case of "probably better not to touch this" given the very big implications!

NeOMakinG replied:
I'm more surprised by the fact that every occurences of `<PrivateRoute />` includes all the public routes

```jsx
    <Switch location={location}>
      <Route {...rest} path='/markets' />
      <Route {...rest} path='/assets' />
      <Route {...rest} path='/rfox' />
      <Route {...rest} path='/explore' />
      <Route {...rest} path='/pools' />
      <Route {...rest} path='/earn' />
      <Route {...rest} path='/buy-crypto' />
      <Route {...rest} path='/assets' />
      <Route {...rest} path='/flags' />
      <Route {...rest} path='/fox' />
      {hasWallet ? <Route {...rest} /> : <Redirect to={to} />}
    </Switch>
    ```

gomesalexandre replied:
soon rambo'd in theaters near you

woodenfurniture reviewed (changes requested):
Code wise very nice. Bye bye 1inch.


# Testing notes

❓ "Connect wallet" appears when opening trade page - is disabling this out of scope?
https://jam.dev/c/d4c012f7-0be6-40ba-a322-b9f4c6d394ce

❌ Zero network fees on thorchain swapper
![image](https://github.com/user-attachments/assets/aa944586-fbf8-4b92-8472-c41c98e6e833)

❌ Zero network fees on portals
![image](https://github.com/user-attachments/assets/d7ce6711-188c-434b-bc1f-8fd4dc821dc7)

❓ Portals slippage % likely wrong
![image](https://github.com/user-attachments/assets/8ddf0a9a-8185-4726-acf3-06887c8dc218)

❓ [Maybe out of scope?] Previous connected wallet balance appears on swapper without wallet connected
![image](https://github.com/user-attachments/assets/7527e727-364a-4704-a20a-560c5a157ee6)

❓ Can't select SOL as a sell asset
https://jam.dev/c/9fba3326-917b-4fb8-af19-0c718cc92a68

❌ Warning acknowledgement appears over the top of "connect wallet" modal
https://jam.dev/c/b5c63bf0-dc25-4095-9a04-7dcd07437dff

❌ Manual receive address entry stuck on after disconnecting wallet
https://jam.dev/c/32eefc42-b793-4353-a9e1-00cfc82eb867

gomesalexandre commented on packages/swapper/src/swappers/ArbitrumBridgeSwapper/utils/constants.ts:
```diff
@@ -7,3 +8,26 @@ export const ARBITRUM_BRIDGE_SUPPORTED_CHAIN_IDS: SupportedChainIds = {
+// https://etherscan.io/tx/0x933c1b625824abeae6f15b4667e22caa3dc14be4929ae62310e2fbf1d39c7d8b
+// https://etherscan.io/tx/0x695090931bb09e60fd378210a0c204ebb01cc137ff00e576074c79a41cc4fa80
+// https://etherscan.io/tx/0xaaef3d2391e4b07f6f0a7524dadde735d4a9822efc2a611e5e8dd5e2708438cc
+export const fallbackEthDepositGasLimit = bn(100_000)
```
IMO 100 is perfect as an OCDly round number since 100 and 100.05k is not going to be a perceptible difference and this is just fallback fees for visual purposes

gomesalexandre commented:
tyvm for that absolute monster of a review @woodenfurniture!
Replied to every single point below, but the tl;dr is except for the fixed slippage bug, all are going to be tackled once we are deeper in implementing the flow and out of current scope, *or* are cantdos that are going to need to be accommodated nicely in the UI. Such a good state we're in when despite not being even started with the actual flow just yet, we already know of so many ramifications/future issues from the first PR :pray:.

In addition, also fixed Cosmos SDK *rates* in 7e965e66bc:

<img width="1128" alt="image" src="https://gist.github.com/user-attachments/assets/ac17e198-379d-46c9-8048-fd5b2ae90e5b">
<img width="1102" alt="image" src="https://gist.github.com/user-attachments/assets/5a9bf3bd-cac7-4512-a3d6-f1d6975b6f2d">


> ❓ "Connect wallet" appears when opening trade page - is disabling this out of scope?

Yesser that's expected and part of the initial flow, using wallet (`hasWallet`) as a quotes vs. rates discriminator for the time being!

Note public swapper route is disabled as part of the flag (https://jam.dev/c/67e2708a-91f5-4611-ac56-67d85e523a91), all we're concerned about regression-testing wise is that users are unable to get into the swapper route without a wallet connected, as well as them not being able to execute rates. 

> ❌ Zero network fees on thorchain swapper

Nice-catch ser, for UTXOs that's also intended with *rates* (flag on, no wallet) for the time being (ish, see below).

The initial intent of this PR when work was started was a *dumb* fetch of rates (as dumb as can be) without a wallet, which for UTXOs meant no network fees.

Though given the latest draft (i.e us always trying and getting rates with fees estimates at input time without a wallet connected), we absolutely want to leverage *some* kind of upstream estimates for UTXOs, though IMO also for a follow-up PR. Started fiddling with abusing THOR `outbound_fee` for BTC (or other UTXOs), though this just doesn't play nicely with our current data structure, and was stuck with `adapter.getFeeData()` (to get `satsPerByte` / `byteCount`) requiring a `pubkey`. Will pick @kaladinlight's brain for next steps here in a follow-up.

> ❌ Zero network fees on portals

This one is not (and is likely never to be) possible with *rates* (flag on, no wallet).

tl;dr is the Portals `/estimate` endpoint - which is the one that shall be used for rates, doesn't do any kind of magic big server heuristics to calculate some kind of EVM fees without an address / without approvals granted: https://paw.pt/jaklXZ2S
We currently have some extremely ugly logic to try out a mix of `/portals` (quote) and `/estimate` endpoint calls, including a "well-funded magic address" with some dummy token for each chain we use as dummy input token to the Portals `/quote` endpoint so we can get some kind of number back, which is more often than not far from accurate. 

I got in touch with Portals very early on when starting the Portals work to request similar magic as e.g Li.Fi 0x and the like do, unfortunately that's not in their plans. Just bugged them again WRT that, but most likely won't happen, if anything not in the near future. Magic API fees estimations for rates should be considered the exception, not the rule meaning we will definitely need a notion of N/A / unknown for fees vs. them being zero'd out atm. 

> 
> ❓ Portals slippage % likely wrong

Big derp! Fixed in ffdd6656b8

<img width="1098" alt="image" src="https://gist.github.com/user-attachments/assets/c9e5e556-6484-48f7-a399-5ba3f927c0d1">

> 
> ❓ [Maybe out of scope?] Previous connected wallet balance appears on swapper without wallet connected

Out of scope for the very initial BIP44 requirements IMO but very valid, the "disconnect connected wallet" flow in swapper is definitely suboptimal atm and will need to be tackled separately

> ❓ Can't select SOL as a sell asset

Can't repro, though added this lil' cleanup guy which may have helped 897910b057

<img width="608" alt="image" src="https://gist.github.com/user-attachments/assets/f4cd560a-ed3d-4862-9f75-4bb001384d9a">

Can't repro, though added this lil' cleanup guy which may have helped 897910b057
 
> ❌ Warning acknowledgement appears over the top of "connect wallet" modal [jam.dev/c/b5c63bf0-dc25-4095-9a04-7dcd07437dff](https://jam.dev/c/b5c63bf0-dc25-4095-9a04-7dcd07437dff)

Also out of scope but definitely something we will want to tackle as the current flow assumes a single modal and `<WarningAcknowledgement />` impl/consumption will need to be revamped to handle pending connect (cc @neomaking)

> ❌ Manual receive address entry stuck on after disconnecting wallet 

Same comment re: "disconnect connected wallet" bugs category, out of scope

woodenfurniture reviewed (approved):
All requested changes actioned or will be addressed in follow-up when possible, as discussed.
